pkg/spdk/engine.go

```diff
@@ -918,129 +913,127 @@ func (e *Engine) ReplicaAddStart(spdkClient *spdkclient.Client, replicaName, rep
+			latestSnapshotName = snapshotName
 		}
-	}()
-
-	// Syncing with the SPDK TGT server only when the engine is running.
-	if e.State != types.InstanceStateRunning {
-		return fmt.Errorf("invalid state %v for engine %s replica %s add finish", e.State, e.Name, replicaName)
 	}
-
-	if e.Frontend != types.FrontendEmpty {
-		return fmt.Errorf("invalid frontend %v for engine %s replica %s add finish", e.Frontend, e.Name, replicaName)
+	if ancestorSnapshotName == "" || latestSnapshotName == "" {
+		return "", nil, fmt.Errorf("cannot find the ancestor snapshot %s or latest snapshot %s from RW replica %s snapshot map during engine %s replica %s add", ancestorSnapshotName, latestSnapshotName, srcReplicaName, e.Name, dstReplicaName)
 	}
+	rebuildingSnapshotList = []*api.Lvol{}
+	rebuildingSnapshotList = getRebuildingSnapshotList(rpcSrcReplica, ancestorSnapshotName, rebuildingSnapshotList)
 
-	if _, exists := e.ReplicaAddressMap[replicaName]; !exists {
-		return fmt.Errorf("replica %s does not exist in engine %s", replicaName, e.Name)
+	// Ask the source replica to expose the newly created snapshot if the source replica and destination replica are not on the same node.
+	externalSnapshotLvolAddress, err := srcReplicaServiceCli.ReplicaRebuildingSrcStart(srcReplicaName, dstReplicaName, dstReplicaAddress, snapshotName)
+	if err != nil {
+		return "", nil, err
 	}
 
-	if e.ReplicaModeMap[replicaName] != types.ModeWO {
-		return fmt.Errorf("invalid mode %s for engine %s replica %s add finish", e.ReplicaModeMap[replicaName], e.Name, replicaName)
+	// The destination replica attaches the source replica exposed snapshot as the external snapshot then create a head based on it.
+	dstHeadLvolAddress, err := dstReplicaServiceCli.ReplicaRebuildingDstStart(dstReplicaName, srcReplicaName, srcReplicaAddress, snapshotName, externalSnapshotLvolAddress, rebuildingSnapshotList)
+	if err != nil {
+		return "", nil, err
 	}
 
-	replicaBdevList := []string{}
-	for rName, rMode := range e.ReplicaModeMap {
-		if rMode == types.ModeRW {
-			replicaBdevList = append(replicaBdevList, e.ReplicaBdevNameMap[rName])
-		}
-		if rName == replicaName {
-			bdevName, err := e.connectReplica(spdkClient, replicaName, replicaAddress)
-			if err != nil {
-				e.log.WithError(err).Errorf("Failed to get bdev from rebuilding replica %s with address %s, will mark it as ERR and error out", replicaName, replicaAddress)
-				e.ReplicaModeMap[replicaName] = types.ModeERR
-				e.ReplicaBdevNameMap[replicaName] = ""
-				updateRequired = true
-				return err
-			}
-			e.ReplicaModeMap[replicaName] = types.ModeRW
-			e.ReplicaBdevNameMap[replicaName] = bdevName
-			replicaBdevList = append(replicaBdevList, bdevName)
-			updateRequired = true
-		}
+	dstHeadLvolBdevName, err := e.connectReplica(spdkClient, dstReplicaName, dstHeadLvolAddress)
+	if err != nil {
+		return "", nil, err
+	}
+	if _, err := spdkClient.BdevRaidGrowBaseBdev(e.Name, dstHeadLvolBdevName); err != nil {
```

Comment:
Also if you put the base bdev in write only mode, some read operations are needed because of the copy on write: if you write on the head lvol over a cluster that is not still allocated, this cluster is copied from the base snapshot and then updated. So, in SPDK, also in you put the base bdev in write only mode, you continue to need a parent for that volume.
For this reason the development to set the base bdev as write only was discarded last year and we started to use the external snapshot, which resolve the above issue.